text/0000-lifecycle-all.md

```diff
@@ -0,0 +1,109 @@
+As members of the community experiment with incorporating CNBs into other platform and CI tools, the difficulty of invoking each lifecycle phase correctly and sequentially draws many to the easier solution of using `pack` to orchestrate the lifecycle. While `pack` can provide a simple way for these developers to implement these integrations, there may be performance drawbacks when compared with using the lifecycle directly. `/cnb/lifecycle/all` will provide an easier interface, lowering the barrier to using the lifecycle without pack, thus enabling platform developers choose between `pack` and `lifecycle` based solely upon the needs of their integration.
+
+# What it is
+[what-it-is]: #what-it-is
+
+### Usage
+
+`/cnb/lifecycle/all <image-name>`
+
+The following flags optional flags can customize behavior:
+
+| Flag            | Required | Env var                | Default                   | Description |
+| --------------- | -------- | ---------------------- | ------------------------- | ------------|
+| `-layers`       | optional | `CNB_LAYERS_DIR`       | `/layers`                 | path to layers |
+| `-platform`     | optional | `CNB_PLATFORM_DIR`     | `/platform`               | path to platform directory |
+| `-run-image`    | optional | `CNB_RUN_IMAGE`        | -                         | reference to run image |
+| `-app`          | optional | `CNB_APP_DIR`          | `/workspace`              | path to app directory |
+| `-cache-dir`    | optional | `CNB_CACHE_DIR`        | -                         | path to cache directory |
+| `-cache-image`  | optional | `CNB_CACHE_IMAGE`      | -                         | cache image tag reference |
+| `-order`        | optional | `CNB_ORDER_PATH`       | `/cnb/order.toml`         | path to order.toml |
+| `-stack`        | optional | `CNB_STACK_PATH`       | `/cnb/stack.toml`         | path to stack.toml |
+| `-launch-cache` | optional | `CNB_LAUNCH_CACHE_DIR` | -                         | path to launch cache directory |
+| `-launcher`     | optional | `CNB_LAUNCHER_PATH`    | `/cnb/lifecycle/launcher` | path to launcher binary |
+| `-buildpacks`   | optional | `CNB_BUILDPACKS_DIR`   | `/cnb/buildpacks`         | path to buildpacks directory |
+| `-daemon`       | optional | `CNB_USE_DAEMON`       | false                     | export to docker daemon |
+| `-uid`          | required | `CNB_USER_ID`          | -                         | UID of user in the stack's build and run images |
+| `-gid`          | required | `CNB_GROUP_ID`         | -                         | GID of user's group in the stack's build and run images |
+| `-version`      | optional | -                      | false                     | show version |
+| `-skip-restore` | optional | -                      | false                     | do not restore metadata from previous image or data from cache
+| `-tag` (multiples allowed) | optional | -           |                           | additional tags to apply to exported image
```

Comment:
What image would the lifecycle analyze for previous layers? Could we add a `previous-image` to separate the tags that we right to from the resulting location?

Reply:
The argument provided as the `<image name>` would be the one that the lifecycle would analyze, no? `-tag` will be used to specify additional write locations.

Reply:
It seems a bit confusing to me that `tags` are not the complete list of locations that the lifecycle will export to especially because <image-name> doesn't have the `tag` name inside of it. 

Additionally, at times it can be useful to explicitly call out the image input to the `analyze` step. For example: kpack could pass in the previous image's digest to the analyzer. You obviously wouldn't export to an image name with the digest.

Reply:
@matthewmcnew I am not opposed to adding an optional flag for the previous image image, so long as it default to the export tag when not supplied.

The rationale behind the single positional argument is that an image tag is the only **required** input. Whereas every other flag configures the build process, the `<image-name>` is the primary thing the process acts on.

Given that we already enforce that all exported tags are tags on the same registry, what if we changed the value passed to `-t` to be a "tag tag" (in the `<registry.com/repo/name:tag`> sense). I am not sure exporting to multiple different repositories in the same registry is really a valid use case. Maybe that would make the usage clearer - only the positional argument gets to define the repository.

Reply:
Love the optional previous image flag.

Reply:
I fear using "tag tag" as the value passed to -t would be confusing because it would be differ from the "tag" terminology in docker and kpack.

Reply:
Could we just reuse the multiple positional arguments like what is in exporter today for multiple tags?

Reply:
A single positional argument and additional `-tag`s seems more honest? The `exporter` currently treats the first of the positional arguments differently than the rest. In this proposed phase the positional argument will be used as the default value for the analyzer image. If one of the tags in special, 1 positional arg + flags seems to indicate that more clearly.